### DISCUSSION TITLE
SSL_CTX_new/SSL_new/SSL_connect memory leaks

### DISCUSSION THREAD

**Discussion by qhdhuowen** [openssl/openssl#23358] [Q&A]
[openssl.txt](https://github.com/openssl/openssl/files/14010502/openssl.txt)
Hi, everyone, like the content in the OpenSSL. txt file, I use openssl to implement tls communication client program, and my client program cannot exit, need to constantly initiate and disconnect links, openssl version is: OpenSSL 1.1.1k FIPS 25 Mar 2021, however, whether I use thread calls, or use process program calls, I find that the memory information occupied by the program is increasing, there are two detection methods, respectively, mtrace to view the stack information, and use pidstat to view the amount of memory occupied by the program. Excuse me, am I using something wrong? Or，there is a memory leak inside openssl. Please help me！

**Comment by t8m** [comment]
I'd suggest using valgrind to find out where the leaks are, if there are any.

**Reply by qhdhuowen** [reply]
[configurecmd.txt](https://github.com/openssl/openssl/files/14016963/configurecmd.txt)
[openssl_trace.log](https://github.com/openssl/openssl/files/14016964/openssl_trace.log)
[openssl_trace_result.log](https://github.com/openssl/openssl/files/14016965/openssl_trace_result.log)
[valgrind_memcheck.log](https://github.com/openssl/openssl/files/14016966/valgrind_memcheck.log)
hello，The valgrind_memcheck.log file is the result file of the valgrind run, and the 512 bytes of direct memory leak is in mtrace, which is normal, I know. Because of business requirements, my program executes in a while(1) loop, and I use ctrl+c to end valgrind's capture memory result. Also, in configurecmd.txt is my openssl compilation and installation instruction; openssl_trace_result.log is the result of capturing memory information using mtrace. Openssl_trace_result. log is the result of mtrace. The execution command is also in the file. For your reference.

**Reply by qhdhuowen** [reply]
When my program is running while(1), use "pidstat -p 'ps -ef | grep". ./openssl_test" | grep -v grep | awk '{print $2}' '-r "command to view the RSS field, you can see that the value is constantly increasing, especially when multithreading, the increase is particularly fast

**Comment by nhorman** [comment]
outside of the mtrace leak, none of these records appear to be real positives.  Each record indicates that the allocation is still reachable via a pointer in the library.  Normally these would be cleaned up when the program terminates normally via return from main or an explicit call to exit().  However, since you terminate the program via SIGINT, atexit handlers aren't run (as it is not normal program termination).  You can fix that by modifying your code to install a signal handler for SIGINT that breaks your while loop, and exits normally.

I've built your supplied code (thank you for that), and run it with -m 20, and the RSS size seems to increase and stabilize quickly, at around 6080kb

**Reply by qhdhuowen** [reply]
<img width="947" alt="openssl_pidstat" src="https://github.com/openssl/openssl/assets/148953339/cb5452b9-2fa8-48d9-990f-41f718f9c233">
Thank you for your reply. I understand what you mean. I will take SIGNAL's suggestion. As shown in the attached screenshot, in my environment, -m 20, RSS information keeps increasing and will not stabilize at a certain fixed value. Also, I used mtrace to monitor malloc/free information and found that several global functions, such as SSL_library_init(), do not consume more memory for multiple loops, which I understand is normal, they save the global information. SSL_CTX_new/SSL_new/SSL_connect three functions, each loop, will increase 0x70 bytes of data do not release, I have used SSL_free/SSL_CTX_free/OPENSSL_thread_stop to release memory information, Why does RSS keep growing? How can this 0x70 bytes of data be released? Looking forward to your reply.

**Comment by qhdhuowen** [comment]
[openssl.txt](https://github.com/openssl/openssl/files/14072304/openssl.txt)
There may be a memory leak related to thread management in openssl. 
For example, run the./openssl_test -m 0 -t 100000 command in openssl.txt. The memory leak cannot be detected using pidstat or mtrace. Therefore, the process in openssl is not leaked. Run the./openssl_test -m 1 -t 100000 command. You can use pidstat to check that the RSS field is increased and that the mtrace information between 1.1 and 1.2 is not leaked. The mtrace information between 3.1 and 3.2 has a memory leak. Therefore, the openssl thread invocation has a memory leak. Memory leakage is independent of SSL_new, SSL_connect, and SSL_ctx_new. Please help to confirm whether this is the case

**Reply by WolfJMZ** [reply]
I am also suspicious that OpenSSL is leaking memory within the context of multi-threaded operations. I am using cURL and OpenSSL statically compiled and embedded and have noticed that the call to SSL_CTX_new() is indeed being flagged as leaking memory by confirming with calls to 
```
	_CrtSetDbgFlag(_CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);
	_CrtSetBreakAlloc(51921);
	_CrtSetBreakAlloc(50996);
	_CrtSetBreakAlloc(50995);
```
in my main method.

![3492d5c361af4e8e535881fada3c24bd](https://github.com/user-attachments/assets/e1601e02-8d28-4632-8f83-fd6c84a77d85)

**Reply by nhorman** [reply]
According to the trace man page, mtrace is not thread safe, and usage in mututhrraded programs leads the missed allocation records

**Reply by WolfJMZ** [reply]
So does that mean it's false positives for memory leaks?

**Reply by nhorman** [reply]
Potentially, yes, though its impossible to tell as your test program doesn't use mtrace in the multithreaded case, nor does it provide a t option as described.

Though, looking at your code you may have some legitimate leaks in your test case, which you may or may not consider to be false positives.  The test as you have written it never exits, so I'm assuming that the only way to terminate it is via SIGINT after you terminate the server.  You have multiple error paths in your test that neglect to free your allocated resources, and dependent on when you hit ctrl-c to terminate the test case, you may fall into any one of those, leading to leaks, because you never called any of the needed free functions when the SIGINT arrived.  Those aren't leaks in openssl however, they are leaks in the test program.

Additionally, you should ignore the RSS values, increasing RSS is not in any way an indication of leaking memory.
